transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go

```diff
@@ -0,0 +1,407 @@
+		)
+
+		BeforeEach(func() {
+			test_config.CleanTestDB(db)
+			repo.SetDB(db)
+			blockOne := rand.Int()
+			blockTwo := blockOne + 1
+			rawTimestampOne := int64(rand.Int31())
+			rawTimestampTwo := rawTimestampOne + 1
+			headerOne = CreateHeaderWithHash(hashOne.String(), rawTimestampOne, blockOne, db)
+			headerTwo = CreateHeaderWithHash(hashTwo.String(), rawTimestampTwo, blockTwo, db)
+			diffID = CreateFakeDiffRecord(db)
+			var parseErr error
+			bidID, parseErr = strconv.Atoi(input.Metadata.Keys[constants.BidId])
+			Expect(parseErr).NotTo(HaveOccurred())
+			var addressErr error
+			addressID, addressErr = shared.GetOrCreateAddress(input.ContractAddress, db)
+			Expect(addressErr).NotTo(HaveOccurred())
+		})
+
+		It("inserts a row for new bid + blockNumber combination", func() {
+			initialBidValues := test_helpers.GetCommonBidStorageValues(0, bidID)
+			test_helpers.InsertValues(db, input.Repository, headerOne, initialBidValues,
+				test_helpers.GetCommonBidMetadatas(strconv.Itoa(bidID)))
+
+			updatedRepoVal, _ := randomBidStorageValue(input.Metadata.Type, input.PackedValueType)
+			err := input.Repository.Create(diffID, headerTwo.Id, input.Metadata, updatedRepoVal)
+			Expect(err).NotTo(HaveOccurred())
+
+			var bidSnapshots []commonBidSnapshot
+			queryErr := db.Select(&bidSnapshots, getBidStateQuery)
+			Expect(queryErr).NotTo(HaveOccurred())
+			Expect(len(bidSnapshots)).To(Equal(2))
+			initialBidValues[input.Metadata.Name] = updatedRepoVal
+			expectedBid := commonBidSnapshotFromValues(bidID, headerTwo.BlockNumber, addressID, headerTwo.Timestamp,
+				initialBidValues)
+			assertBidSnapshot(bidSnapshots[1], expectedBid, headerTwo.BlockNumber)
+			assertSingleField(bidSnapshots[1], expectedBid, input.ColumnName)
+		})
+
+		It("updates record if bid record already exists for block", func() {
+			initialBidValues := test_helpers.GetCommonBidStorageValues(0, bidID)
+			test_helpers.InsertValues(db, repo, headerOne, initialBidValues,
+				test_helpers.GetCommonBidMetadatas(strconv.Itoa(bidID)))
+
+			updatedRepoVal, _ := randomBidStorageValue(input.Metadata.Type, input.PackedValueType)
+			err := input.Repository.Create(diffID, headerOne.Id, input.Metadata, updatedRepoVal)
+			Expect(err).NotTo(HaveOccurred())
+
+			var bidSnapshots []commonBidSnapshot
+			queryErr := db.Select(&bidSnapshots, getBidStateQuery)
+			Expect(queryErr).NotTo(HaveOccurred())
+			Expect(len(bidSnapshots)).To(Equal(1))
+			initialBidValues[input.Metadata.Name] = updatedRepoVal
+			expectedBid := commonBidSnapshotFromValues(bidID, headerTwo.BlockNumber, addressID, headerOne.Timestamp,
+				initialBidValues)
+			assertBidSnapshot(bidSnapshots[0], expectedBid, headerOne.BlockNumber)
+			assertSingleField(bidSnapshots[0], expectedBid, input.ColumnName)
+		})
+	})
+}
+
+func randomBidStorageValue(valueType vdbStorage.ValueType, packedValueType vdbStorage.ValueType) (interface{}, string) {
```

Comment:
This function was my best effort at keeping packed-value repository logic out of the body of the tests. It returns 2 values: the first is what the `repo.Create` function wants to accept, and the second is the string value that will actually be inserted into the trigger-table, which we can use in assertions or raw SQL inserts. Usually these 2 values will be the same, except in the case of packed values.

Reply:
Might be worth exploring how much clutter we'd create by extracting a separate suite/function for packed values? Totally on board with the de-duplicating approach, but wondering if this is a place where there's a need for some kind of polymorphism to handle important differences between the data structures/avoid returning vals that are only useful to a subset of callers

Reply:
Yeah that's a good call. I guess in this case I was so preoccupied with whether or not I _could_ combine all the tests suites into one, that I didn’t stop to think if I _should_. ;)

Reply:
The other thing I considered in order to avoid this problem entirely, is to create a utility function that handles inserting storage records, rather than relying on the repository. My thinking is that the trigger tests have nothing to do with the repositories (especially the 'packed values' portion of the repo logic), so the fact that we even need to think about packed values in the trigger tests feels like sort of a code smell to me. Maybe it would be worth decoupling these tests from the repository.